src/hotspot/cpu/riscv/assembler_riscv.hpp

```diff
@@ -2813,24 +2812,35 @@ enum Nf {
 // --------------------------
 // Unconditional branch instructions
 // --------------------------
-#define INSN(NAME)                                                                           \
-  void NAME(Register Rd, Register Rs, const int32_t offset) {                                \
-    /* jalr -> c.jr/c.jalr */                                                                \
-    if (do_compress() && (offset == 0 && Rs != x0)) {                                        \
-      if (Rd == x1) {                                                                        \
-        c_jalr(Rs);                                                                          \
-        return;                                                                              \
-      } else if (Rd == x0) {                                                                 \
-        c_jr(Rs);                                                                            \
-        return;                                                                              \
-      }                                                                                      \
-    }                                                                                        \
-    _jalr(Rd, Rs, offset);                                                                   \
+ protected:
+  // All calls and jumps must go via MASM.
+  void jalr(Register Rd, Register Rs, const int32_t offset) {
+    /* jalr -> c.jr/c.jalr */
+    if (do_compress() && (offset == 0 && Rs != x0)) {
+      if (Rd == x1) {
+        c_jalr(Rs);
+        return;
+      } else if (Rd == x0) {
+        c_jr(Rs);
+        return;
+      }
+    }
+    _jalr(Rd, Rs, offset);
   }
 
-  INSN(jalr);
+  void jal(Register Rd, const int32_t offset) {
+    /* jal -> c.j, note c.jal is RV32C only */
+    if (do_compress() &&
+        Rd == x0 &&
+        is_simm12(offset) && ((offset % 2) == 0)) {
+      c_j(offset);
```

Comment:
Is RV32C-only instructions usable for our case? Or will this if block be test covered?

Reply:
There are two instructions using the CJ format: c.j and c.jal.
- c.j is already in our assembler.
- c.jal is not in our assembler and is RV32C only.

000000000000064e <c_j>:
     64e: 11 a0        	j	0x652 <c_j+0x4>  // 0xa011 is c.j 4
     650: 2a 85        	mv	a0, a0
     652: 82 80        	ret

Works fine on both VF2 and qemu-rv64.

The comment was just saying that we can't try to map to c.jal since it's RV32C.

The instruction c.j have the same test coverage as before (as we already have it in assembler).
The toogle to C I think is untested as it only can be generated for short backwards branches in non-relocated code.

As the comment is obvious confusing, suggestions to change it?

Reply:
Uh, I see. I think I mis-read the comment. I am OK with the current shape.